### PR TITLE
Added option to not render empty literals on nested json objects

### PR DESCRIPTION
**Pull request by johnkors**
- Nested objects of null value were rendered as empty json literals. Added option to remove these.

### CODE REVIEW

**Comment by 304NotModified** [comment]
Thanks! We review it!

File: src/NLog/Layouts/JsonLayout.cs [hunk 1]
```diff
@@ -68,6 +69,11 @@ public JsonLayout()
         public bool SuppressSpaces { get; set; }
 
         /// <summary>
+        /// Gets or sets the option to render the empty object value {}
+        /// </summary>
+        public bool RenderEmptyLiteral { get; set; }
```
Comment (304NotModified) [inline src/NLog/Layouts/JsonLayout.cs]: I was thinking of 'RenderEmptyObject`? What do you think?

File: src/NLog/Layouts/JsonLayout.cs [hunk 2]
```diff
@@ -53,6 +53,7 @@ public class JsonLayout : Layout
         public JsonLayout()
         {
             this.Attributes = new List<JsonAttribute>();
+            this.RenderEmptyLiteral = true;
```
Comment (304NotModified) [inline src/NLog/Layouts/JsonLayout.cs]: I think it's wise to keep the behavior backwards-compatible (so `false` as default)

File: src/NLog/Layouts/JsonLayout.cs [hunk 3]
```diff
@@ -119,7 +125,17 @@ protected override string GetFormattedMessage(LogEventInfo logEvent)
             AppendIf(!this.SuppressSpaces, sb, " ");
             sb.Append("}");
 
-            return sb.ToString();
+
+            var result = sb.ToString();
+
+            if (!RenderEmptyLiteral)
+            {
+                if (result.Replace(" ", "") == "{}")
```
Comment (304NotModified) [inline src/NLog/Layouts/JsonLayout.cs]: This works, but it is a bit hacky ;)

I was thinking, postpone adding the `{` and `}` (eg. https://github.com/johnkors/NLog/blob/c2d5366d574c155b34591c96c31aa8c05c67730b/src/NLog/Layouts/JsonLayout.cs#L85) and add them if the object isn't empty.

File: src/NLog/Layouts/JsonLayout.cs [hunk 4]
```diff
@@ -119,7 +125,17 @@ protected override string GetFormattedMessage(LogEventInfo logEvent)
...
+                    return null;
```
Comment (304NotModified) [inline src/NLog/Layouts/JsonLayout.cs]: string.empty please :)

File: src/NLog/Layouts/JsonLayout.cs (hunk 2 above)
Comment (johnkors) [inline src/NLog/Layouts/JsonLayout.cs]: That's why i set it to true by default. The behaviour was to render empty literals by default, and I wanted an option of removing them, so one has to opt out by setting it to false.

File: src/NLog/Layouts/JsonLayout.cs (hunk 3 above)
Comment (johnkors) [inline src/NLog/Layouts/JsonLayout.cs]: Yeah, I agree :)

File: src/NLog/Layouts/JsonLayout.cs (hunk 1 above)
Comment (johnkors) [inline src/NLog/Layouts/JsonLayout.cs]: Yeah, better naming!

File: src/NLog/Layouts/JsonLayout.cs [hunk 5]
```diff
@@ -116,10 +120,18 @@ protected override string GetFormattedMessage(LogEventInfo logEvent)
                 }
             }
 
-            AppendIf(!this.SuppressSpaces, sb, " ");
-            sb.Append("}");
+            var result = sb.ToString();
+
+            if (string.IsNullOrEmpty(result.Trim()) && !RenderEmptyObject)
```
Comment (304NotModified) [inline src/NLog/Layouts/JsonLayout.cs]: It's maybe even better to keep a boolean value tracking the "empty state"? 

Less String operations are better (e.g. trim and string comparisons)

**Comment by 304NotModified** [comment]
One not, let me know what you think

File: src/NLog/Layouts/JsonLayout.cs (hunk 5 above)
Comment (johnkors) [inline src/NLog/Layouts/JsonLayout.cs]: Ah, okay! Yeah, not used to thinking performance up front :) Thanks. Will push an update.

Comment (304NotModified) [inline src/NLog/Layouts/JsonLayout.cs]: Cool!

Well pretty important stuff to most of the users ;)

**Comment by 304NotModified** [comment]
perfect!

**Comment by 304NotModified** [comment]
Added on wiki https://github.com/NLog/NLog/wiki/JsonLayout

**Comment by 304NotModified** [comment]
4.3.7 is online: https://www.nuget.org/packages/NLog/4.3.7

**Pull request merged**